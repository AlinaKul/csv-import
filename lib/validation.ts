export type Error = {
  code: string
  column: string
  line: number
}

export type ValidationOptions = {
  fromLine?: number
}

export type ValidationResult = {
  isValid: boolean
  errors: Error[]
}

export class ValidationContext {
  private currentRow: number
  private errors: Error[]

  constructor(options: ValidationOptions = {}) {
    this.currentRow = options.fromLine ? options.fromLine : 1
    this.errors = []
  }

  public nextRow(): void {
    this.currentRow++
  }

  addError(column: string, errorCode: string): void {
    this.errors.push({
      column,
      line: this.currentRow,
      code: errorCode,
    })
  }

  getResult(): ValidationResult {
    return {
      isValid: this.errors.length === 0,
      errors: this.errors,
    }
  }
}
