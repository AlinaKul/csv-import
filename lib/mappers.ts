import { ValidationContext } from './validation'

export enum ErrorCodes {
  MISSING_MANDATORY_FIELD = 'MISSING_MANDATORY_FIELD',
  INVALID_DATE_FORMAT = 'INVALID_DATE_FORMAT',
  INVALID_BOOLEAN_FORMAT = 'INVALID_BOOLEAN_FORMAT',
  INVALID_INTEGER_FORMAT = 'INVALID_INTEGER_FORMAT',
  INVALID_FLOAT_FORMAT = 'INVALID_FLOAT_FORMAT',
}

export type Mapper<T> = (
  column: string,
  input: string,
  validationContext: ValidationContext,
  defaultValue: T | undefined
) => T

export function toOptionalString(
  column: string,
  input: string,
  validationContext: ValidationContext,
  defaultValue: string | undefined
): string | undefined {
  if (input.length) {
    return input
  }
  return defaultValue
}

export function toMandatoryString(
  column: string,
  input: string,
  validationContext: ValidationContext
): string | undefined {
  if (input.length) {
    return input
  }
  validationContext.addError(column, ErrorCodes.MISSING_MANDATORY_FIELD)
  return undefined
}

export function toMandatoryInteger(
  column: string,
  input: string,
  validationContext: ValidationContext
): number | undefined {
  if (input.length) {
    const value = parseInt(input)
    if (!Number.isNaN(value)) {
      return value
    }
    validationContext.addError(column, ErrorCodes.INVALID_INTEGER_FORMAT)
    return undefined
  }
  validationContext.addError(column, ErrorCodes.MISSING_MANDATORY_FIELD)
  return undefined
}

export function toOptionalInteger(
  column: string,
  input: string,
  validationContext: ValidationContext,
  defaultValue: number | undefined
): number | undefined {
  if (input.length) {
    const value = parseInt(input)
    if (!Number.isNaN(value)) {
      return value
    }
    validationContext.addError(column, ErrorCodes.INVALID_INTEGER_FORMAT)
    return undefined
  }
  return defaultValue
}

export function toMandatoryFloat(
  column: string,
  input: string,
  validationContext: ValidationContext
): number | undefined {
  if (input.length) {
    const value = parseFloat(input)
    if (!Number.isNaN(value)) {
      return value
    }
    validationContext.addError(column, ErrorCodes.INVALID_FLOAT_FORMAT)
    return undefined
  }
  validationContext.addError(column, ErrorCodes.MISSING_MANDATORY_FIELD)
  return undefined
}

export function toOptionalFloat(
  column: string,
  input: string,
  validationContext: ValidationContext,
  defaultValue: number | undefined
): number | undefined {
  if (input.length) {
    const value = parseFloat(input)
    if (!Number.isNaN(value)) {
      return value
    }
    validationContext.addError(column, ErrorCodes.INVALID_FLOAT_FORMAT)
    return undefined
  }
  return defaultValue
}

export function toMandatoryBoolean(
  column: string,
  input: string,
  validationContext: ValidationContext,
  defaultValue: boolean | undefined
): boolean | undefined {
  if (input.length) {
    if (input.toLowerCase() === 'true' || input === '1') {
      return true
    }
    if (input.toLowerCase() === 'false' || input === '0') {
      return false
    }
    validationContext.addError(column, ErrorCodes.INVALID_BOOLEAN_FORMAT)
    return undefined
  }
  return defaultValue
}

export function toOptionalBoolean(
  column: string,
  input: string,
  validationContext: ValidationContext,
  defaultValue: '' | boolean | undefined
): boolean | undefined | '' {
  if (input.length) {
    if (input == '1') {
      return true
    }
    if (!input) {
      return false
    }
    validationContext.addError(column, ErrorCodes.INVALID_BOOLEAN_FORMAT)
    return undefined
  }
  return defaultValue
}

export function toOptionalDate(
  column: string,
  input: string,
  validationContext: ValidationContext,
  defaultValue: Date | undefined
): Date | undefined {
  if (input.length) {
    const value = new Date(input)
    if (isNaN(value.getTime())) {
      validationContext.addError(column, ErrorCodes.INVALID_DATE_FORMAT)
      return undefined
    }
    return value
  }
  return defaultValue
}

export function toMandatoryDate(
  column: string,
  input: string,
  validationContext: ValidationContext
): Date | undefined {
  if (input.length) {
    const value = new Date(input)
    if (isNaN(value.getTime())) {
      validationContext.addError(column, ErrorCodes.INVALID_DATE_FORMAT)
      return undefined
    }
    return value
  }
  validationContext.addError(column, ErrorCodes.MISSING_MANDATORY_FIELD)
  return undefined
}
