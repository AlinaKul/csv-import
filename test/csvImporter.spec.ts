import { resolve } from 'path'
import { getFileStream } from './utils/fileUtils'
import { ColumnDefinitions, loadFromCsv } from '../lib/csvImporter'
import {
  toMandatoryBoolean,
  toMandatoryDate,
  toMandatoryFloat,
  toMandatoryInteger,
  toMandatoryString,
  toOptionalBoolean,
  toOptionalDate,
  toOptionalFloat,
  toOptionalInteger,
  toOptionalString,
} from '../lib/mappers'

describe('csvImporter', () => {
  describe('loadEntities', () => {
    it('correctly loads large file', async () => {
      const filePath = resolve(__dirname, 'data', 'stop_times.csv')
      const stream = getFileStream(filePath)
      const columnConfig: ColumnDefinitions = [
        {
          column: 'trip_id',
          entityField: 'tripId',
          mapper: toMandatoryInteger,
        },
        {
          column: 'arrival_time',
          entityField: 'arrival_time',
          mapper: toOptionalString,
        },
        {
          column: 'departure_time',
          entityField: 'departure_time',
          mapper: toMandatoryString,
        },
        {
          column: 'stop_id',
          entityField: 'stop_id',
          mapper: toMandatoryInteger,
        },
        {
          column: 'stop_sequence',
          entityField: 'stop_sequence',
          mapper: toMandatoryInteger,
        },
        {
          column: 'pickup_type',
          entityField: 'pickup_type',
          mapper: toMandatoryInteger,
        },
        {
          column: 'drop_off_type',
          entityField: 'drop_off_type',
          mapper: toMandatoryInteger,
        },
        {
          column: 'destinationTraveled',
          entityField: 'destinationTraveled',
          mapper: toOptionalInteger,
          defaultValue: 0,
        },
        {
          column: 'priority',
          entityField: 'priority',
          mapper: toOptionalBoolean,
          defaultValue: false,
        },
      ]
      const result = await loadFromCsv(stream, columnConfig, {
        fromLine: 2,
      })
      console.log(`Import time: ${result.importTime}`)
      expect(result.validationResult).toMatchSnapshot()
      expect(result.data.length).toEqual(512781)
      expect(result.data[0]).toMatchSnapshot()
      expect(result.data[8785]).toMatchSnapshot()
      expect(result.data[112780]).toMatchSnapshot()
      expect(result.data[212780]).toMatchSnapshot()
      expect(result.data[312780]).toMatchSnapshot()
      expect(result.data[512760]).toMatchSnapshot()
      expect(result.data[512780]).toMatchSnapshot()
    }, 90000)
    it('correctly loads large file stopsWithErrors', async () => {
      const filePath = resolve(__dirname, 'data', 'stopsWithErrors.csv')
      const stream = getFileStream(filePath)
      const columnConfig: ColumnDefinitions = [
        {
          column: 'stopId',
          entityField: 'stopId',
          mapper: toMandatoryString,
        },
        {
          column: 'code',
          entityField: 'code',
          mapper: toOptionalInteger,
        },
        {
          column: 'name',
          entityField: 'name',
          mapper: toMandatoryString,
        },
        {
          column: 'description',
          entityField: 'description',
          mapper: toOptionalString,
        },
        {
          column: 'latitude',
          entityField: 'latitude',
          mapper: toMandatoryFloat,
        },
        {
          column: 'longitude',
          entityField: 'longitude',
          mapper: toMandatoryFloat,
        },
        {
          column: 'url',
          entityField: 'url',
          mapper: toOptionalString,
        },
        {
          column: 'stopTypeId',
          entityField: 'stopTypeId',
          mapper: toMandatoryString,
        },
        {
          column: 'hasWifi',
          entityField: 'hasWifi',
          mapper: toOptionalBoolean,
        },
        {
          column: 'managerId',
          entityField: 'managerId',
          mapper: toOptionalString,
        },

        {
          column: 'geom',
          entityField: 'geom',
          mapper: toOptionalString,
        },
        {
          column: 'isActive',
          entityField: 'isActive',
          mapper: toMandatoryBoolean,
        },
      ]
      const result = await loadFromCsv(stream, columnConfig, {
        fromLine: 2,
      })
      console.log(result.data, `Import time: ${result.importTime}`)
      // expect(result.validationResult).toMatchSnapshot()
      expect(result.data.length).toEqual(6)
      // expect(result.data).toMatchSnapshot()
      // expect(result.data[8785]).toMatchSnapshot()
      // expect(result.data[112780]).toMatchSnapshot()
      // expect(result.data[212780]).toMatchSnapshot()
      // expect(result.data[312780]).toMatchSnapshot()
      // expect(result.data[512760]).toMatchSnapshot()
      // expect(result.data[512780]).toMatchSnapshot()
    }, 90000)
    it('correctly loads medium file', async () => {
      const filePath = resolve(__dirname, 'data', 'stop_times_mid.csv')
      const stream = getFileStream(filePath)
      const columnConfig: ColumnDefinitions = [
        {
          column: 'trip_id',
          entityField: 'tripId',
          mapper: toMandatoryInteger,
        },
        {
          column: 'arrival_time',
          entityField: 'arrival_time',
          mapper: toOptionalString,
        },
        {
          column: 'departure_time',
          entityField: 'departure_time',
          mapper: toMandatoryString,
        },
        {
          column: 'stop_id',
          entityField: 'stop_id',
          mapper: toMandatoryInteger,
        },
        {
          column: 'stop_sequence',
          entityField: 'stop_sequence',
          mapper: toMandatoryInteger,
        },
        {
          column: 'pickup_type',
          entityField: 'pickup_type',
          mapper: toMandatoryInteger,
        },
        {
          column: 'drop_off_type',
          entityField: 'drop_off_type',
          mapper: toMandatoryInteger,
        },
        {
          column: 'destinationTraveled',
          entityField: 'destinationTraveled',
          mapper: toOptionalFloat,
          defaultValue: 0.0,
        },
        {
          column: 'priority',
          entityField: 'priority',
          mapper: toOptionalBoolean,
          defaultValue: true,
        },
      ]
      const result = await loadFromCsv(stream, columnConfig, {
        fromLine: 2,
      })
      console.log(`Import time: ${result.importTime}`)
      expect(result.data.length).toEqual(5040)
      expect(result.data[0]).toMatchSnapshot()
      expect(result.data[1020]).toMatchSnapshot()
      expect(result.data[5030]).toMatchSnapshot()
      expect(result.data[5039]).toMatchSnapshot()
    }, 90000)

    it('correctly returns errors', async () => {
      const filePath = resolve(__dirname, 'data', 'stop_times_errors.csv')
      const stream = getFileStream(filePath)
      const columnConfig: ColumnDefinitions = [
        {
          column: 'trip_id',
          entityField: 'tripId',
          mapper: toMandatoryString,
        },
        {
          column: 'arrival_time',
          entityField: 'arrival_time',
          mapper: toOptionalDate,
        },
        {
          column: 'departure_time',
          entityField: 'departure_time',
          mapper: toMandatoryDate,
        },
        {
          column: 'stop_id',
          entityField: 'stop_id',
          mapper: toMandatoryInteger,
        },
        {
          column: 'stop_sequence',
          entityField: 'stop_sequence',
          mapper: toMandatoryInteger,
        },
        {
          column: 'pickup_type',
          entityField: 'pickup_type',
          mapper: toMandatoryInteger,
        },
        {
          column: 'drop_off_type',
          entityField: 'drop_off_type',
          mapper: toMandatoryInteger,
        },
        {
          column: 'destinationTraveled',
          entityField: 'destinationTraveled',
          mapper: toOptionalString,
        },
        {
          column: 'priority',
          entityField: 'priority',
          mapper: toOptionalBoolean,
        },
      ]
      const result = await loadFromCsv(stream, columnConfig, {
        fromLine: 2,
      })
      console.log(`Import time: ${result.importTime}`)
      expect(result.validationResult).toMatchSnapshot()
      expect(result.data.length).toEqual(7)
      expect(result.data).toMatchSnapshot()
    })
  })
})
