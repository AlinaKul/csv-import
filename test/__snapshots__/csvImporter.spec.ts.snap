// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`csvImporter loadEntities correctly loads large file 1`] = `
Object {
  "errors": Array [],
  "isValid": true,
}
`;

exports[`csvImporter loadEntities correctly loads large file 2`] = `
Object {
  "arrival_time": "04:49:00",
  "departure_time": "04:50:00",
  "destinationTraveled": 0,
  "drop_off_type": 0,
  "pickup_type": 0,
  "priority": false,
  "stop_id": 8262,
  "stop_sequence": 1,
  "tripId": 1,
}
`;

exports[`csvImporter loadEntities correctly loads large file 3`] = `
Object {
  "arrival_time": "15:23:00",
  "departure_time": "15:23:00",
  "destinationTraveled": 0,
  "drop_off_type": 0,
  "pickup_type": 0,
  "priority": false,
  "stop_id": 5836,
  "stop_sequence": 6,
  "tripId": 716,
}
`;

exports[`csvImporter loadEntities correctly loads large file 4`] = `
Object {
  "arrival_time": "11:38:00",
  "departure_time": "11:38:00",
  "destinationTraveled": 0,
  "drop_off_type": 0,
  "pickup_type": 0,
  "priority": false,
  "stop_id": 6083,
  "stop_sequence": 9,
  "tripId": 6375,
}
`;

exports[`csvImporter loadEntities correctly loads large file 5`] = `
Object {
  "arrival_time": "11:22:00",
  "departure_time": "11:22:00",
  "destinationTraveled": 0,
  "drop_off_type": 0,
  "pickup_type": 0,
  "priority": false,
  "stop_id": 16581,
  "stop_sequence": 12,
  "tripId": 11523,
}
`;

exports[`csvImporter loadEntities correctly loads large file 6`] = `
Object {
  "arrival_time": "22:26:00",
  "departure_time": "22:26:00",
  "destinationTraveled": 0,
  "drop_off_type": 0,
  "pickup_type": 0,
  "priority": false,
  "stop_id": 6497,
  "stop_sequence": 31,
  "tripId": 16273,
}
`;

exports[`csvImporter loadEntities correctly loads large file 7`] = `
Object {
  "arrival_time": "19:22:00",
  "departure_time": "19:22:00",
  "destinationTraveled": 0,
  "drop_off_type": 0,
  "pickup_type": 0,
  "priority": false,
  "stop_id": 5086,
  "stop_sequence": 3,
  "tripId": 25564,
}
`;

exports[`csvImporter loadEntities correctly loads large file 8`] = `
Object {
  "arrival_time": "06:14:00",
  "departure_time": "06:14:00",
  "destinationTraveled": 0,
  "drop_off_type": 0,
  "pickup_type": 0,
  "priority": false,
  "stop_id": 5505,
  "stop_sequence": 8,
  "tripId": 25566,
}
`;

exports[`csvImporter loadEntities correctly loads medium file 1`] = `
Object {
  "arrival_time": "04:49:00",
  "departure_time": "04:50:00",
  "destinationTraveled": 0,
  "drop_off_type": 0,
  "pickup_type": 0,
  "priority": false,
  "stop_id": 8262,
  "stop_sequence": 1,
  "tripId": 1,
}
`;

exports[`csvImporter loadEntities correctly loads medium file 2`] = `
Object {
  "arrival_time": "07:26:00",
  "departure_time": "07:26:00",
  "destinationTraveled": 0,
  "drop_off_type": 0,
  "pickup_type": 0,
  "priority": true,
  "stop_id": 7573,
  "stop_sequence": 8,
  "tripId": 127,
}
`;

exports[`csvImporter loadEntities correctly loads medium file 3`] = `
Object {
  "arrival_time": "13:22:00",
  "departure_time": "13:22:00",
  "destinationTraveled": 0,
  "drop_off_type": 0,
  "pickup_type": 0,
  "priority": true,
  "stop_id": 5156,
  "stop_sequence": 1,
  "tripId": 466,
}
`;

exports[`csvImporter loadEntities correctly loads medium file 4`] = `
Object {
  "arrival_time": "13:49:00",
  "departure_time": "13:49:00",
  "destinationTraveled": 0,
  "drop_off_type": 0,
  "pickup_type": 0,
  "priority": true,
  "stop_id": 6487,
  "stop_sequence": 10,
  "tripId": 466,
}
`;

exports[`csvImporter loadEntities correctly returns errors 1`] = `
Object {
  "errors": Array [
    Object {
      "code": "INVALID_DATE_FORMAT",
      "column": "arrival_time",
      "line": 3,
    },
    Object {
      "code": "INVALID_INTEGER_FORMAT",
      "column": "pickup_type",
      "line": 3,
    },
    Object {
      "code": "MISSING_MANDATORY_FIELD",
      "column": "drop_off_type",
      "line": 4,
    },
    Object {
      "code": "INVALID_BOOLEAN_FORMAT",
      "column": "priority",
      "line": 4,
    },
    Object {
      "code": "INVALID_DATE_FORMAT",
      "column": "arrival_time",
      "line": 5,
    },
    Object {
      "code": "MISSING_MANDATORY_FIELD",
      "column": "departure_time",
      "line": 6,
    },
    Object {
      "code": "MISSING_MANDATORY_FIELD",
      "column": "trip_id",
      "line": 8,
    },
    Object {
      "code": "INVALID_DATE_FORMAT",
      "column": "departure_time",
      "line": 8,
    },
  ],
  "isValid": false,
}
`;

exports[`csvImporter loadEntities correctly returns errors 2`] = `
Array [
  Object {
    "arrival_time": 2011-10-10T00:00:00.000Z,
    "departure_time": 2011-10-11T00:00:00.000Z,
    "destinationTraveled": undefined,
    "drop_off_type": 0,
    "pickup_type": 0,
    "priority": false,
    "stop_id": 8262,
    "stop_sequence": 1,
    "tripId": "a",
  },
  Object {
    "arrival_time": undefined,
    "departure_time": 2011-10-11T00:00:00.000Z,
    "destinationTraveled": undefined,
    "drop_off_type": 0,
    "pickup_type": undefined,
    "priority": undefined,
    "stop_id": 6650,
    "stop_sequence": 2,
    "tripId": "b",
  },
  Object {
    "arrival_time": 2011-10-10T00:00:00.000Z,
    "departure_time": 2011-10-11T00:00:00.000Z,
    "destinationTraveled": undefined,
    "drop_off_type": undefined,
    "pickup_type": 0,
    "priority": undefined,
    "stop_id": 7851,
    "stop_sequence": 3,
    "tripId": "c",
  },
  Object {
    "arrival_time": undefined,
    "departure_time": 2011-10-11T00:00:00.000Z,
    "destinationTraveled": undefined,
    "drop_off_type": 0,
    "pickup_type": 0,
    "priority": undefined,
    "stop_id": 5173,
    "stop_sequence": 4,
    "tripId": "d",
  },
  Object {
    "arrival_time": 2011-10-10T00:00:00.000Z,
    "departure_time": undefined,
    "destinationTraveled": undefined,
    "drop_off_type": 0,
    "pickup_type": 0,
    "priority": true,
    "stop_id": 5175,
    "stop_sequence": 5,
    "tripId": "e",
  },
  Object {
    "arrival_time": 2011-10-10T00:00:00.000Z,
    "departure_time": 2011-10-11T00:00:00.000Z,
    "destinationTraveled": undefined,
    "drop_off_type": 0,
    "pickup_type": 0,
    "priority": undefined,
    "stop_id": 10359,
    "stop_sequence": 6,
    "tripId": "f",
  },
  Object {
    "arrival_time": 2011-10-10T00:00:00.000Z,
    "departure_time": undefined,
    "destinationTraveled": undefined,
    "drop_off_type": 0,
    "pickup_type": 0,
    "priority": undefined,
    "stop_id": 9738,
    "stop_sequence": 7,
    "tripId": undefined,
  },
]
`;
